File: Backend/Microservices/Commodity/Eefa.Commodity.Application/Commands/CommodityCategoryProperty/Create/CreateCommodityCategoryPropertyCommand.cs
using AutoMapper;
using Eefa.Common.CommandQuery;
using Eefa.Common;
using Eefa.Common.Data;
using MediatR;
using System.Threading;
using System.Threading.Tasks;
using Eefa.Commodity.Application.Queries.Commodity;
using Eefa.Commodity.Application.Commands.CommodityCategoryPropertyItem.Create;
using System.Collections.Generic;

namespace Eefa.Commodity.Application.Commands.CommodityCategoryProperty.Create
{
    public class CreateCommodityCategoryPropertyCommand : CommandBase, IRequest<ServiceResult<CommodityCategoryPropertyModel>>, IMapFrom<CreateCommodityCategoryPropertyCommand>, ICommand
    {

        /// <summary>
        /// کد والد
        /// </summary>
        public int? ParentId { get; set; }

        /// <summary>
        /// کد گروه
        /// </summary>
        public int? CategoryId { get; set; }

        /// <summary>
        /// کد سطح
        /// </summary>
        public string LevelCode { get; set; } = default!;

        /// <summary>
        /// نام اختصاصی
        /// </summary>
        public string UniqueName { get; set; } = default!;

        /// <summary>
        /// عنوان
        /// </summary>
        public string Title { get; set; } = default!;

        /// <summary>
        /// واحد اندازه گیری
        /// </summary>
        public int? MeasureId { get; set; }

        /// <summary>
        /// قوانین حاکم بر مولفه
        /// </summary>
        public int? PropertyTypeBaseId { get; set; }

        /// <summary>
        /// ترتیب نمایش
        /// </summary>
        public int OrderIndex { get; set; } = default!;


        public List<CreateCommodityCategoryPropertyItemCommand> Items { get; set; }
        public void Mapping(Profile profile)
        {
            profile.CreateMap<CreateCommodityCategoryPropertyCommand, Data.Entities.CommodityCategoryProperty>()
                .ForMember(x => x.CommodityCategoryPropertyItems, opt => opt.MapFrom(x => x.Items));
        }
    }

    public class CreateCommodityCategoryPropertyCommandHandler : IRequestHandler<CreateCommodityCategoryPropertyCommand, ServiceResult<CommodityCategoryPropertyModel>>
    {
        private readonly IRepository<Data.Entities.CommodityCategoryProperty> _repository;
        private readonly IMapper _mapper;

        public CreateCommodityCategoryPropertyCommandHandler(IRepository<Data.Entities.CommodityCategoryProperty> repository, IMapper mapper)
        {
            _mapper = mapper;
            _repository = repository;
        }


        public async Task<ServiceResult<CommodityCategoryPropertyModel>> Handle(CreateCommodityCategoryPropertyCommand request, CancellationToken cancellationToken)
        {
            var commodityCategoryProperty = _mapper.Map<Data.Entities.CommodityCategoryProperty>(request);

            var entity = _repository.Insert(commodityCategoryProperty);

            await _repository.SaveChangesAsync(cancellationToken);

            return ServiceResult<CommodityCategoryPropertyModel>.Success(
                _mapper.Map<CommodityCategoryPropertyModel>(entity));
        }
    }
}

File: Frontend/angular-app/src/app/modules/commodity/repositories/commodity-category-property/commands/create-commodity-category-property-command.ts
import {Inject} from "@angular/core";
import {ServiceResult} from "../../../../../core/models/service-result";
import {IRequest, IRequestHandler} from "../../../../../core/services/mediator/interfaces";
import {HttpService} from "../../../../../core/services/http/http.service";
import {ApplicationError} from "../../../../../core/exceptions/application-error";
import {MediatorHandler} from "../../../../../core/services/mediator/decorator";
import {CommodityCategoryProperty} from "../../../entities/commodity-category-property";
import {NotificationService} from "../../../../../shared/services/notification/notification.service";
import {ValidationRule} from "../../../../../core/validation/validation-rule";
import {HttpRequest} from "../../../../../core/services/http/http-request";
import {
  CreateCommodityCategoryPropertyItemCommand
} from "../../commodity-category-property-item/commands/create-commodity-category-property-item-command";

export class CreateCommodityCategoryPropertyCommand extends IRequest<CreateCommodityCategoryPropertyCommand, CommodityCategoryProperty> {

  public parentId: number | undefined = undefined;
  public categoryId: number | undefined = undefined;
  public levelCode: string | undefined = undefined;
  public uniqueName: string | undefined = undefined;
  public title: string | undefined = undefined;
  public measureId: number | undefined = undefined;
  public propertyTypeBaseId: number | undefined = undefined;
  public orderIndex: number | undefined = undefined;
  public items: CreateCommodityCategoryPropertyItemCommand[] = [];

  constructor() {
    super();
  }

  mapFrom(entity: CommodityCategoryProperty): CreateCommodityCategoryPropertyCommand {
    throw new ApplicationError(CreateCommodityCategoryPropertyCommand.name, this.mapTo.name, 'Not Implemented Yet')
  }

  mapTo(): CommodityCategoryProperty {
    throw new ApplicationError(CreateCommodityCategoryPropertyCommand.name, this.mapTo.name, 'Not Implemented Yet')
  }

  get url(): string {
    return "/commodity/commodityCategoryProperty/add";
  }

  get validationRules(): ValidationRule[] {
    return [];
  }
}

@MediatorHandler(CreateCommodityCategoryPropertyCommandHandler.name)
export class CreateCommodityCategoryPropertyCommandHandler implements IRequestHandler<CreateCommodityCategoryPropertyCommand, CommodityCategoryProperty> {
  constructor(
    @Inject(HttpService) private _httpService: HttpService,
    @Inject(NotificationService) private _notificationService: NotificationService,
  ) {
  }

  async Handle(request: CreateCommodityCategoryPropertyCommand): Promise<CommodityCategoryProperty> {
    let httpRequest: HttpRequest<CreateCommodityCategoryPropertyCommand> = new HttpRequest<CreateCommodityCategoryPropertyCommand>(request.url, request);
    this._notificationService.isLoader = true;
    return await this._httpService.Post<CreateCommodityCategoryPropertyCommand, ServiceResult<CommodityCategoryProperty>>(httpRequest).toPromise().then(response => {
      this._notificationService.showSuccessMessage()
      return response.objResult
    }).finally(() => {
      this._notificationService.isLoader = false;
    })
  }
}
